### ISSUE TITLE
net-snmp is not changing authentication and privacy protocol.

### ISSUE THREAD

**Issue by mdshoaibumer** [net-snmp/net-snmp#470]
I'm using the net-snmp library (version 5.7.1) in a C++ programming under windows. I have a Web-Frontend where a user can choose a SNMP-Version and configure it. SNMPv1 and SNMPv2 are working just fine, but I got some issues with SNMPv3.

When I start the C++ backend and enter all needed SNMPv3 credentials correctly, everything is working fine and the device is reachable. If I change for example the Authentication Protocol from MD5 to SHA, but leave the rest of the credentials the same, I would expect that the device is not reachable anymore. In real it stays reachable. After restarting the backend the device is (as expected) not reachable anymore with the same settings.

My conclusion after testing was, that net-snmp seems to cache the selected authentication and privacy protocol for one user name. This can be seen very good at Test 2. The first time I use an user name with a specific protocol I get the expected result. After changing the protocol a different result is expected, but I get still the same result as before.

At the end some information how the SNMP-calls are made:

There is a class called SNMPWrapper, which handels the whole SNMP-communication.
Inside the constructor I call init_snmp() to init net-snmp.
From the outside I can call only get(), set() and walk(). Every time one of these methods is called, a new SNMP-Session is created (First I create a new Session with snmp_sess_init(), than I set up the things needed and finally I open the session with snmp_sess_open().
After I made the request and received my answer I close the session with snmp_sess_close().

int main(int argc, char** argv) {
struct snmp_session session, session1, *ss, *ss1;
struct snmp_pdu *pdu, *pdu1;
struct snmp_pdu *response, *response1;

oid anOID[MAX_OID_LEN];
size_t anOID_len = MAX_OID_LEN;

struct variable_list *vars;
int status, status1;

init_snmp("snmpapp");

const char* user = "md5";
string authpw = "123123123";
string privpw = "";
string ipString = "192.168.15.32";

char ip[16];
memset(&ip, 0, sizeof (ip));
ipString.copy(ip, sizeof (ip) - 1, 0);

/*
 * First request: AuthProto is MD5, no PrivProto is used. The snmp-get
 * request is successful
 */
snmp_sess_init(&session); /* set up defaults */
session.peername = ip;
session.version = SNMP_VERSION_3;

/* set the SNMPv3 user name */
session.securityName = strdup(user);
session.securityNameLen = strlen(session.securityName);

// set the authentication method to MD5     
session.securityLevel = SNMP_SEC_LEVEL_AUTHNOPRIV;

session.securityAuthProto = usmHMACMD5AuthProtocol;
session.securityAuthProtoLen = USM_AUTH_PROTO_MD5_LEN;
session.securityAuthKeyLen = USM_AUTH_KU_LEN;;

if (generate_Ku(session.securityAuthProto,
        session.securityAuthProtoLen,
        (u_char *) authpw.c_str(), strlen(authpw.c_str()),
        session.securityAuthKey,
        &session.securityAuthKeyLen) != SNMPERR_SUCCESS) {
    //if code reaches here, the creation of the security key was not successful

}

cout << "SecurityAuthProto - session: " << session.securityAuthProto[9] << " / SecurityAuthKey - session: " << session.securityAuthKey << endl;

ss = snmp_open(&session); /* establish the session */

if (!ss) {
    cout << "Couldn't open session1 correctly";
    exit(2);
}

cout << "SecurityAuthProto - ss: " << ss->securityAuthProto[9] << " / SecurityAuthKey - ss: " << ss->securityAuthKey << endl;

//send message
pdu = snmp_pdu_create(SNMP_MSG_GET);
read_objid(".1.3.6.1.2.1.1.1.0", anOID, &anOID_len);
snmp_add_null_var(pdu, anOID, anOID_len);
status = snmp_synch_response(ss, pdu, &response);

/*
 * Process the response.
 */
if (status == STAT_SUCCESS && response->errstat == SNMP_ERR_NOERROR) {
    cout << "SNMP-read success" << endl;
} else {
    cout << "SNMP-read fail" << endl;
}

if (response)
    snmp_free_pdu(response);
if (!snmp_close(ss))
    cout << "Snmp closing failed" << endl;

/*
 * Second request: Only the authProto is changed from MD5 to SHA1. I expect,
 * that the snmp-get fails, but it still succeeds.
 */

snmp_sess_init(&session1);
session1.peername = ip;
session1.version = SNMP_VERSION_3;

/* set the SNMPv3 user name */
session1.securityName = strdup(user);
session1.securityNameLen = strlen(session1.securityName);

// set the authentication method to SHA1 
session1.securityLevel = SNMP_SEC_LEVEL_AUTHNOPRIV;

session1.securityAuthProto = usmHMACSHA1AuthProtocol;
session1.securityAuthProtoLen = USM_AUTH_PROTO_SHA_LEN;
session1.securityAuthKeyLen = USM_AUTH_KU_LEN;

if (generate_Ku(session1.securityAuthProto,
        session1.securityAuthProtoLen,
        (u_char *) authpw.c_str(), strlen(authpw.c_str()),
        session1.securityAuthKey,
        &session1.securityAuthKeyLen) != SNMPERR_SUCCESS) {
    //if code reaches here, the creation of the security key was not successful
}

cout << "SecurityAuthProto - session1: " << session1.securityAuthProto[9] << " / SecurityAuthKey - session1: " << session1.securityAuthKey << endl;

ss1 = snmp_open(&session1); /* establish the session */

if (!ss1) {
    cout << "Couldn't open session1 correctly";
    exit(2);
}

cout << "SecurityAuthProto - ss1: " << ss1->securityAuthProto[9] << " / SecurityAuthKey - ss1: " << ss1->securityAuthKey << endl;

//send message
pdu1 = snmp_pdu_create(SNMP_MSG_GET);
read_objid(".1.3.6.1.2.1.1.1.0", anOID, &anOID_len);
snmp_add_null_var(pdu1, anOID, anOID_len);
status1 = snmp_synch_response(ss1, pdu1, &response1);

/*
 * Process the response.
 */
if (status1 == STAT_SUCCESS && response1->errstat == SNMP_ERR_NOERROR) {
    cout << "SNMP-read success" << endl;
} else {
    cout << "SNMP-read fail" << endl;
}

if (response1)
    snmp_free_pdu(response1);
snmp_close(ss1);

return 0;
}

Kindly help me on this how to overcome this issue. I found solution on other forum.
net-snmp caches for every EngineId (device) the users. If there is an existing user for an engineID and you try to open a new session with this user, net-snmp will use the cached one.
1)  To clear the list with cached users.
2) To update the password of that securityname.

But my Question is when mutithreads are running we may face the issue of updating usm table and removing a user from usm table  during runtime at great cost. Kindly help us to overcome this issue .

**Comment by Flow86** [comment]
Hi, did you find any solution for this? I'm experiencing the same issue.

**Comment by teusbenschop** [comment]
The solution for this is to clear the Net-SNMP internal cache with user credentials.

```
   usmUser* actual_user = usm_get_userList();
    while (actual_user != NULL) {
        usmUser* row_pointer = actual_user;
        usm_remove_user(actual_user);
        actual_user = row_pointer->next;
    }
```

See also for more information this link: https://stackoverflow.com/questions/18380435/net-snmp-is-not-changing-auth-and-priv-protocol-correctly

**Comment by Flow86** [comment]
@teusbenschop 
thanks, thats actually working, but: In a non-threaded environment that is a solution, but not for a multi threaded one, or if you have multiple connection at the same time. 
Deleting all the users in the usm list will delete any authentication/session information, which "kills" any open snmp v3 connections.

**Comment by teusbenschop** [comment]
Yeah, likely true. 

Your situation may be different, but in our backend we have some kind of error checking, that if a session is killed, it restarts this session again (and then calculates new authentication information).

**Comment by teusbenschop** [comment]
> [...] thats actually working, but only in a non-threaded environment - not for a multi threaded one. [...]

This is correct. For the multi-threaded use of this, to prevent crashes, we used a mutex and locked this during the time that this USM user list is accessed. It is accessed while clearing it in the example above, and secondly it is access while generating the password hash:

```
generate_Ku(session.securityAuthProto, session.securityAuthProtoLen, password, strlen(password), session.securityAuthKey, &session.securityAuthKeyLen);
```

Locking the mutex in both cases fixed all the crashes in the multi-threaded environment in our situation.

**Comment by mdshoaibumer** [comment]
Is this resolved if i upgrade to the latest version of NetSNMP?
